modules/EvseManager/IECStateMachine.cpp

```diff
@@ -0,0 +1,451 @@
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            timeout_state_c1.stop();
+            call_allow_power_on_bsp(false);
+        }
+        break;
+
+    case RawCPState::A:
+        if (last_cp_state != RawCPState::A) {
+            pwm_running = false;
+            r_bsp->call_pwm_off();
+            ev_simplified_mode = false;
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 2.1 Unplug at state Bx (or any other
+        // state) Table A.6: Sequence 2.2 Unplug at state Cx, Dx
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::Disabled) {
+            events.push(CPEvent::CarRequestedStopPower);
+            events.push(CPEvent::CarUnplugged);
+        }
+        break;
+
+    case RawCPState::B:
+        // Table A.6: Sequence 7 EV stops charging
+        // Table A.6: Sequence 8.2 EV supply equipment
+        // responds to EV opens S2 (w/o PWM)
+
+        if (last_cp_state != RawCPState::A && last_cp_state != RawCPState::B) {
+
+            events.push(CPEvent::CarRequestedStopPower);
+            // Need to switch off according to Table A.6 Sequence 8.1
+            // within 100ms
+            call_allow_power_on_bsp(false);
+            timeout_state_c1.stop();
+        }
+
+        // Table A.6: Sequence 1.1 Plug-in
+        if (last_cp_state == RawCPState::A || last_cp_state == RawCPState::Disabled) {
+            events.push(CPEvent::CarPluggedIn);
+            ev_simplified_mode = false;
+        }
+
+        if (last_cp_state == RawCPState::E || last_cp_state == RawCPState::F) {
+            // Triggers SLAC start
+            events.push(CPEvent::EFtoBCD);
+        }
+        break;
+
+    case RawCPState::D:
+        // If state D is not supported, signal an error event here and switch off.
+        if (not has_ventilation) {
+            call_allow_power_on_bsp(false);
+            events.push(CPEvent::ErrorVentilationNotAvailable);
+            timeout_state_c1.stop();
+            break;
+        }
+        // no break, intended fall through: If we support state D it is handled the same way as state C
+    case RawCPState::C:
+        // Table A.6: Sequence 1.2 Plug-in
+        if (last_cp_state == RawCPState::A) {
+            events.push(CPEvent::CarPluggedIn);
+            EVLOG_info << "Detected simplified mode.";
+            ev_simplified_mode = true;
+        }
+        if (last_cp_state == RawCPState::B) {
+            events.push(CPEvent::CarRequestedPower);
+        }
+
+        if (!pwm_running && last_pwm_running) { // X2->C1
+                                                // Table A.6 Sequence 10.2: EV does not stop drawing power
+                                                // even if PWM stops. Stop within 6 seconds (E.g. Kona1!)
+            timeout_state_c1.start(std::chrono::seconds(6));
+        }
+
+        if (timeout_state_c1.reached()) {
```

Comment:
Will this only be triggered if we receive another event from the bsp? Can't we use the timer to actually call `call_allow_power_on_bsp` directly?

Reply:
We would still need to check if we are in state C before etc, I think it is more readable to have it in the state machine